weave/state/graph_client.py

```diff
@@ -0,0 +1,42 @@
+import threading
+from typing import TYPE_CHECKING, Optional
+
+from weave.errors import WeaveInitError
+from weave.legacy import context_state
+
+if TYPE_CHECKING:
+    from weave.weave_client import WeaveClient
+
+_global_graph_client: Optional["WeaveClient"] = None
+lock = threading.Lock()
+
+
+def set_graph_client_global(client: Optional["WeaveClient"]) -> None:
+    global _global_graph_client
+
+    # These outer guards are to avoid expensive lock acquisition
+    if client is not None and _global_graph_client is None:
+        with lock:
+            if _global_graph_client is None:
+                _global_graph_client = client
+
+    elif client is None and _global_graph_client is not None:
+        with lock:
+            if _global_graph_client is not None:
+                _global_graph_client = client
```

Comment:
Most operations will set the client globally, which means it should "just work" in threaded cases